apps/aecore/src/aec_chain_state.erl

```diff
@@ -0,0 +1,458 @@
+
+total_difficulty_at_hash(Hash, State) ->
+    total_difficulty_at_hash(Hash, 0, State).
+
+total_difficulty_at_hash(Hash, Acc, State) ->
+    case blocks_db_find(Hash, State) of
+        {ok, Node} ->
+            NewAcc = Acc + node_difficulty(Node),
+            PrevHash = prev_hash(Node),
+            case node_is_genesis(Node) of
+                true -> {ok, NewAcc};
+                false -> total_difficulty_at_hash(PrevHash, NewAcc, State)
+            end;
+        error ->
+            {error, not_rooted}
+    end.
+
+%%%-------------------------------------------------------------------
+%%% Handling the state trees
+%%%-------------------------------------------------------------------
+
+add_state_tree_to_block(Block, Hash, State) ->
+    case state_db_find(Hash, State) of
+        {ok, Trees} -> {ok, aec_blocks:set_trees(Block, Trees)};
+        error -> error
+    end.
+
+%%%-------------------------------------------------------------------
+%%% Chain operations
+%%%-------------------------------------------------------------------
+
+internal_insert(Node, State) ->
+    case blocks_db_find(Node#node.hash, State) of
+        error -> check_update_after_insert(Node, blocks_db_put(Node, State));
+        {ok, Node} -> State;
+        {ok, Old} when Node#node.type =:= block, Old#node.type =:= header ->
+            check_update_after_insert(Node, blocks_db_put(Node, State));
+        {ok, Old} when Node#node.type =:= Old#node.type ->
+            internal_error({same_key_different_content, Node, Old});
+        {ok, Old} when Node#node.type =:= header, Old#node.type =:= block ->
+            case export_header(Old) =:= export_header(Node) of
+                true  -> State;
+                false -> internal_error({different_header_existing_block, Old, Node})
+            end
+    end.
+
+check_update_after_insert(Node, State) ->
+    case determine_chain_relation(Node, State) of
+        off_chain -> State;
+        in_chain -> update_state_tree(Node, State);
+        {fork, ForkNode} ->
+            NewTopNodes = find_new_header_top_from_node(ForkNode, State),
+            NewTopNode = determine_new_header_top(NewTopNodes, State),
+            State1 = set_top_header_hash(NewTopNode#node.hash, State),
+            update_state_tree(ForkNode, State1);
+        new_top ->
+            [NewTopNode] = find_new_header_top_from_node(Node, State),
+            State1 = set_top_header_hash(NewTopNode#node.hash, State),
+            update_state_tree(Node, State1)
+    end.
+
+determine_chain_relation(Node, State) ->
+    Height   = node_height(Node),
+    Hash     = Node#node.hash,
+    PrevHash = prev_hash(Node),
+    assert_previous_height(Height, PrevHash, State),
+    case get_top_header_hash(State) of
+        PrevHash -> new_top;
+        undefined when Height > 0 ->
+            off_chain; %% No proper chain yet.
+        undefined when Height =:= 0 ->
+            new_top; %% A genesis block. TODO: Should this be checked?
```

Comment:
I think it should be checked, yes.

Reply:
This is only when there was no genesis block before.